src/datetime.rs

```diff
@@ -206,31 +206,25 @@ impl<U: Unit> fmt::Debug for Timedelta<U> {
 
 struct TypeDescriptors {
     npy_type: NPY_TYPES,
-    dtypes: UnsafeCell<Option<FxHashMap<NPY_DATETIMEUNIT, Py<PyArrayDescr>>>>,
+    #[allow(clippy::type_complexity)]
+    dtypes: GILProtected<UnsafeCell<Option<FxHashMap<NPY_DATETIMEUNIT, Py<PyArrayDescr>>>>>,
 }
 
-unsafe impl Sync for TypeDescriptors {}
-
 impl TypeDescriptors {
     /// `npy_type` must be either `NPY_DATETIME` or `NPY_TIMEDELTA`.
     const unsafe fn new(npy_type: NPY_TYPES) -> Self {
         Self {
             npy_type,
-            dtypes: UnsafeCell::new(None),
+            dtypes: GILProtected::new(UnsafeCell::new(None)),
         }
     }
 
-    #[allow(clippy::mut_from_ref)]
-    unsafe fn get(&self) -> &mut FxHashMap<NPY_DATETIMEUNIT, Py<PyArrayDescr>> {
-        (*self.dtypes.get()).get_or_insert_with(Default::default)
-    }
-
     #[allow(clippy::wrong_self_convention)]
     fn from_unit<'py>(&'py self, py: Python<'py>, unit: NPY_DATETIMEUNIT) -> &'py PyArrayDescr {
-        // SAFETY: We hold the GIL and we do not call into user code which might re-enter.
-        let dtypes = unsafe { self.get() };
+        // SAFETY: We do not call into user code which might re-enter.
+        let dtypes = unsafe { &mut *self.dtypes.get(py).get() };
```

Comment:
`GILProtected` is admittedly not pulling a lot of weight here but at least the `unsafe impl Sync` is gone.

I did try to make this fully safe using `RefCell` instead of `UnsafeCell`, but I would have to return `RefMut<PyArrayDescr>` instead of `&'py PyArrayDescr` then which does work when the call chain reaches the implementation of `Element::get_dtype`.

Reply:
Actually, I am just incompetent: I need `.clone().into_ref(py)` instead of `.as_ref(py)`.

This does imply a "useless" reference count increase, but this appears necessary as the compiler understandably cannot see that the `Py<PyArrayDescr>` are essentially immortal (as we never remove entries from the hash table) and hence we do need to defend against someone dropping that `Py<PyArrayDescr>` while the `&'py PyArrayDescr` is still alive.